crates/cheatnet/src/execution/syscall_interceptor.rs

```diff
@@ -0,0 +1,206 @@
+use cairo_felt::Felt252;
+use cairo_vm::hint_processor::hint_processor_definition::HintReference;
+use cairo_vm::serde::deserialize_program::ApTracking;
+use cairo_vm::types::exec_scope::ExecutionScopes;
+use cairo_vm::vm::errors::hint_errors::HintError;
+use cairo_vm::vm::errors::vm_errors::VirtualMachineError;
+use cairo_vm::vm::runners::cairo_runner::RunResources;
+use cairo_vm::vm::vm_core::VirtualMachine;
+use std::any::Any;
+use std::collections::HashMap;
+
+pub struct ExecuteHintRequest<'a> {
+    pub vm: &'a mut VirtualMachine,
+    pub exec_scopes: &'a mut ExecutionScopes,
+    pub hint_data: &'a Box<dyn Any>,
+    pub constants: &'a HashMap<String, Felt252>,
+}
+
+pub trait ChainableHintProcessor {
+    fn get_child(&self) -> Option<&dyn HintProcessorLogicInterceptor>;
+    fn get_child_mut(&mut self) -> Option<&mut dyn HintProcessorLogicInterceptor>;
+}
+
+pub trait HintExecutionInterceptor: ChainableHintProcessor {
+    fn intercept_execute_hint(
+        &mut self,
+        execute_hint_request: &mut ExecuteHintRequest,
+    ) -> Option<Result<(), HintError>>;
+
+    fn next_execute_hint(
+        &mut self,
+        execute_hint_request: &mut ExecuteHintRequest,
+    ) -> Option<Result<(), HintError>> {
+        if let Some(result) = self.intercept_execute_hint(execute_hint_request) {
+            return Some(result);
+        }
+        if let Some(child) = self.get_child_mut() {
+            child.next_execute_hint(execute_hint_request)
+        } else {
+            None
+        }
+    }
+
+    fn execute_hint_chain(
+        &mut self,
+        vm: &mut VirtualMachine,
+        exec_scopes: &mut ExecutionScopes,
+        hint_data: &Box<dyn Any>,
+        constants: &HashMap<String, Felt252>,
+    ) -> Result<(), HintError> {
+        if let Some(result) = self.next_execute_hint(&mut ExecuteHintRequest {
+            vm,
+            exec_scopes,
+            hint_data,
+            constants,
+        }) {
+            return result;
+        }
+        unreachable!("execute_hint failed to terminate with a result")
```

Comment:
Why this branching is possible on the type level? If we separated Extension and main syscall handler we could gurantee on the type level that any hint will be handled and we could avoid unreachable branches.

Reply:
I doubt that we'll be able to avoid this kind of panic/unreachable branch, since we want all the handlers to handle a subset of the hints

Reply:
Syscall Handlers handles the full set, so each hint should be eventually handled by the core handler.